### DISCUSSION TITLE
Usability ideas

### DISCUSSION THREAD

**Discussion by Yitaarli** [TobiasM95/WoW-Talent-Tree-Manager#16] [Ideas]
Hey,
This is my first time using GitHub so apologies if I've put this in the wrong place but this didn't feel like it belonged in the issues section.

First of all thanks for making this program. In particular, I'm interested in using the talent tree editor to create (or amend) existing trees to suggest improvements to the developers and also just to pass the time/provoke discussion in the community. I have, tried to editing & creating talent trees and wanted to suggest a list of potential ideas.

I'm not a developer so I don't know how feasible these suggestions are. I've read through the controls but apologies in advance if I have missed something, I haven't had a lot of time to play with this program thus far.

1. Copy and pasting nodes - Being able to select a node and copy & paste it would be really useful if you want to edit an existing tree by moving a node that already exists. I would suggest that it should copy all of the properties of that node except for the children/parents as the node is likely to be moved anyway. You could implement it so that you can select entire portions of the talent tree (e.g. 4 nodes and the links between them) which would make moving entire sections of a tree without changing their fundamental structure a lot easier.

2. Inserting existing nodes - Kind of the same as the first one but it would be easier if you could start a new tree from scratch but insert a node from an existing tree directly into a new tree without having to open the existing tree in another window and then copy and paste it or by insert an empty node and changing the details manually using the existing tool.

3. Swapping nodes - Being able to swap the position and links of two nodes by simply selecting them and using one click or keybind would be much easier than manually going through and removing and re-adding all the parent/children links manually.

4. Different keybinds for linking nodes - The left alt/shift system to add/remove links works but I think another button should be used instead of shift because that button is already used to show the name boxes thus obscuring what you're doing while clicking. Perhaps using P for add/remove parent and C for add/remove child would be easier than alt+click or shift+alt+click? I note that you also had an auto-linking tool. I haven't personally tried this but it might be easier to have a system where you can select multiple nodes and click a button/keybind to automatically link those nodes only. 

5. Dynamically update drop-down boxes by entering letters - If I want to find the tooltip/name for an existing spell, I should be able to just start typing letters and the drop-down box should auto-scroll to the section that starts with those letters. This is similar to how drop-down boxes in a lot of forms work and would speed up having to scroll through the (really long) list of spells in TTM. This is a common feature on database software and in online application forms and is a massive time saver.

6. I would suggest that changes to a talent happen automatically rather than having to scroll to the bottom of the side window and click update each time you change it.

7. In general, having functions incorporated into the toolbar (with keybind included for user information) will allow users to more easily identify everything the program is capable of. See the attached picture from google sheets.

![Program_Example](https://user-images.githubusercontent.com/112441138/187314395-2b18cc58-b3d8-45e6-b72f-69453c986651.png)

To re-iterate, I don't know how feasible any of the above is or what time scale it could be implemented over but I think it would be an improvement over the existing version. 

I'd be interested to hear your thoughts on the above.

EDIT: Grammar/spelling mistakes. I'm writing this far too late at night...

**Comment by Yitaarli** [comment]
Oh and I totally forgot to add, adding a standard redo (Ctrl+Y) and undo (Ctrl+Z) function would be super useful!

**Comment by TobiasM95** [comment]
Thanks for your detailed feedback. It looks to me that most of these things are about the tree editor. There wasn't a major overhaul planned yet but it might be a good idea to make another pass over features and functionality. Some features you mention would need a non-minor UI change and would therefore be a bit of work and could take a while. Nonetheless many of them are good ideas and worth to put in.

Additionally, I'd like to adress each point individually:
1. Copy & paste would be a useful feature indeed. I'm not sure why you'd need a copy function to specifically move a node but in any case c&p is noted as a future feature.
1b. You also mentioned multi talent selection. This is also a reasonable request. I decided against it in the past but might revisit that request in the future.
2. This should be covered by copy & paste
3. If multi talent selection is in, this should be trivial
4. Good catch, the shift hotkey to show talent names was added much later than the hotkeys for linking nodes so I didn't notice that clash. I should rework some hotkeys or maybe implement a new hotkey system to better manage hotkeys.
5. This is not an easy one with the provided library functionality of my GUI library. I'll experiment a bit and see if I can get something to run that's nicer to use. For now, the filter text boxes are a (clunky but otherwise usable) alternative
6. I don't want to do this cause there are complex checks going on when you press update that need to be dealt with in an explizit manner if something goes wrong (e.g. cycles in the tree, etc.) This will stay for now.
7. Not every function can have a pressable toolbar button (for example quick insert connections since you need to hover another talent) but for the ones where it's possible I can extend the toolbar functionality.
8. redo/undo are extremely complex and would require a very big overhaul of the complete inner workings of the editor. While I recognize that this feature would be very nice I have to concede for now due to the complexity this feature would entail.

So in summary, many good ideas which I noted for a future tree editor pass which might come in 1.3.9 or 1.4.0. Thanks again for the feedback, much appreciated.

**Reply by Yitaarli** [reply]
Thanks very much for the response. I agree with everything you said and I look forward to the changes. Thanks again :)
